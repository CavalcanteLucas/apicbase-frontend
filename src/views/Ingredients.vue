<template>
  <div id="ingredients">
    <div class="container">
      <div class="col-12 mx-auto">
        <div class="row">
          <div class="col-4 mx-auto">
            <div class="input-group mb-3">
              <span class="input-group-text" id="search-ingredient"
                ><img src="../assets/img/search.svg" alt="Search"
              /></span>
              <input
                type="text"
                id="searchField"
                class="form-control"
                v-model="searchQuery"
              />
              <button
                class="btn btn-outline-secondary"
                type="button"
                id="clearSearchBtn"
                @click="clearSearchField"
              >
                <img src="../assets/img/backspace.svg" alt="Clear Search" />
              </button>
            </div>
          </div>
        </div>

        <div class="row justify-content-center" style="padding-top: 1rem;">
          <div class="col-md-10 offset-md-1">
            <form id="ingredientForm" @submit.prevent="submitForm">
              <div class="row g-1">
                <div class="col-md-1">
                  <input
                    type="number"
                    class="form-control"
                    id="inputArticleNumber"
                    placeholder="#"
                    v-model="ingredient.article_number"
                    min="1"
                    required="required"
                  />
                </div>
                <div class="col-md-3">
                  <input
                    type="text"
                    class="form-control"
                    id="inputName"
                    placeholder="Name"
                    aria-label="Name"
                    v-model="ingredient.name"
                    required="required"
                  />
                </div>
                <div class="col-md-2">
                  <input
                    type="number"
                    class="form-control"
                    id="inputAmount"
                    placeholder="Amount"
                    v-model="ingredient.amount"
                    min=".01"
                    step=".01"
                    required="required"
                  />
                </div>
                <div class="col-md-1">
                  <select
                    class="form-select"
                    aria-label="Select unit"
                    id="inputUnit"
                    v-model="ingredient.unit"
                    required="required"
                  >
                    <option value="unit">Unit</option>
                    <option value="gram">gram</option>
                    <option value="kilogram">kilogram</option>
                    <option value="liter">liter</option>
                    <option value="centiliter">centiliter</option>
                  </select>
                </div>
                <div class="col-md-2">
                  <div class="input-group mb-3">
                    <span class="input-group-text" id="inputCostPerAmount"
                      >&euro;</span
                    >
                    <input
                      type="number"
                      class="form-control"
                      id="inputCostPerAmount"
                      placeholder="Cost"
                      v-model="ingredient.cost_per_amount"
                      min=".01"
                      step=".01"
                      required="required"
                    />
                  </div>
                </div>
                <div class="col-md-3">
                  <button
                    type="button"
                    id="clearBtn"
                    v-on:click="clearIngredientForm()"
                    class="btn btn-outline-secondary float-start"
                  >
                    <img src="../assets/img/backspace.svg" alt="Clear" />
                  </button>
                  <button
                    type="submit"
                    id="submitBtn"
                    class="btn btn-primary float-start"
                    style="margin-left: 0.25rem;"
                  >
                    Save
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>

        <p class="error" v-if="!isResponseOK">
          Something went wrong. Please try again..
        </p>

        <div class="row" style="padding-top: 2rem;">
          <div class="col-12 mx-auto">
            <table
              class="table table-striped table-hovering"
              aria-label="Ingredients table"
            >
              <thead>
                <th scope="col"><strong>#</strong></th>
                <th scope="col">Name</th>
                <th scope="col">Amount</th>
                <th scope="col">Unit</th>
                <th scope="col">Cost</th>
              </thead>
              <tbody>
                <tr
                  v-for="ingredient in resultQuery"
                  :key="ingredient.id"
                  @click="$data.ingredient = ingredient"
                >
                  <td class="align-middle">
                    <strong>{{ ingredient.article_number }}</strong>
                  </td>
                  <td class="align-middle">{{ ingredient.name }}</td>
                  <td class="align-middle">{{ ingredient.amount }}</td>
                  <td class="align-middle">{{ ingredient.unit }}</td>
                  <td class="align-middle">
                    &euro;{{ ingredient.cost_per_amount }}
                  </td>
                  <td class="align-middle">
                    <button
                      type="button"
                      id="deleteBtn"
                      class="btn btn-sm btn-danger"
                      @click="deleteIngredient(ingredient)"
                    >
                      <img src="../assets/img/trash.svg" alt="Delete" />
                    </button>
                    <confirm-dialogue ref="confirmDialogue"></confirm-dialogue>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>
<script>
import ConfirmDialogue from "../components/ConfirmDialogue.vue";

export default {
  name: "Ingredients",
  components: { ConfirmDialogue },
  data() {
    return {
      searchQuery: null,
      ingredient: {
        article_number: null,
        name: null,
        cost_per_amount: null,
        amount: null,
        unit: "gram",
      },
      ingredients: [],
      isResponseOK: true,
    };
  },
  async created() {
    await this.getIngredients();
  },
  methods: {
    submitForm() {
      if (this.ingredient.id === undefined) {
        this.createIngredient();
      } else {
        this.editIngredient();
      }
    },
    async getIngredients() {
      let response = await fetch("http://localhost:8000/api/ingredients/");
      this.ingredients = await response.json();
    },
    async createIngredient() {
      await this.getIngredients();
      let response = await fetch("http://localhost:8000/api/ingredients/", {
        method: "post",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(this.ingredient),
      });
      this.isResponseOK = response.ok;
      await this.getIngredients();
    },
    async editIngredient() {
      await this.getIngredients();
      let response = await fetch(
        `http://localhost:8000/api/ingredients/${this.ingredient.id}/`,
        {
          method: "put",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(this.ingredient),
        }
      );
      this.isResponseOK = response.ok;
      await this.getIngredients();
    },
    async deleteIngredient(ingredient) {
      await this.getIngredients();
      const ok = await this.$refs.confirmDialogue[0].show();
      if (ok) {
        await fetch(`http://localhost:8000/api/ingredients/${ingredient.id}/`, {
          method: "delete",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(this.ingredient),
        });
      }
      await this.getIngredients();
    },
    clearIngredientForm() {
      this.ingredient = {
        article_number: null,
        name: null,
        cost_per_amount: null,
        amount: null,
        unit: "gram",
      };
      this.isResponseOK = true;
    },
    clearSearchField() {
      this.searchQuery = "";
    },
  },
  computed: {
    resultQuery() {
      if (this.searchQuery) {
        return this.ingredients.filter((ingredient) => {
          return this.searchQuery
            .toLowerCase()
            .split(" ")
            .every((item) => {
              return (
                ingredient.name.toLowerCase().includes(item) ||
                ingredient.article_number == item
              );
            });
        });
      } else {
        return this.ingredients;
      }
    },
  },
};
</script>

<style></style>
