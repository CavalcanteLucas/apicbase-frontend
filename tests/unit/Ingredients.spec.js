import { mount, config } from "@vue/test-utils";
import Ingredients from "../../src/views/Ingredients.vue";

import { columnData } from "../utils.js";

config.showDeprecationWarnings = false;

describe("ingredients view", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(Ingredients, {
      methods: {
        getIngredients: () => {
          // do nothing
        },
      },
    });
  });

  it("renders", () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("has data", () => {
    expect(typeof Ingredients.data).toBe("function");
  });

  it("is a Vue instance", () => {
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("renders the correct table headers", () => {
    expect(wrapper.html()).toContain("#");
    expect(wrapper.html()).toContain("Name");
    expect(wrapper.html()).toContain("Cost");
    expect(wrapper.html()).toContain("Amount");
    expect(wrapper.html()).toContain("Unit");
  });

  it("has five input fields, and a search field", () => {
    expect(wrapper.contains('input[id="inputArticleNumber"]')).toBe(true);
    expect(wrapper.contains('input[id="inputName"]')).toBe(true);
    expect(wrapper.contains('input[id="inputCostPerAmount"]')).toBe(true);
    expect(wrapper.contains('input[id="inputAmount"]')).toBe(true);
    expect(wrapper.contains('select[id="inputUnit"]')).toBe(true);
    expect(wrapper.contains('input[id="searchField"]')).toBe(true);
  });

  it("has clear inputs button, a clear search button, and a submit button", () => {
    expect(wrapper.contains('button[id="clearBtn"]')).toBe(true);
    expect(wrapper.contains('button[id="submitBtn"]')).toBe(true);
    expect(wrapper.contains('button[id="clearSearchBtn"]')).toBe(true);
  });

  it("input fields load data", () => {
    expect(wrapper.vm.$data.ingredient).toEqual({
      article_number: null,
      name: null,
      cost_per_amount: null,
      amount: null,
      unit: "gram",
    });

    wrapper.find('input[id="inputArticleNumber"]').setValue("999");
    wrapper.find('input[id="inputName"]').setValue("carrot-test");
    wrapper.find('input[id="inputCostPerAmount"]').setValue("1");
    wrapper.find('input[id="inputAmount"]').setValue("500");
    wrapper.find('select[id="inputUnit"]').setValue("gram");

    expect(wrapper.find('input[id="inputArticleNumber"]').element.value).toBe(
      "999"
    );
    expect(wrapper.find('input[id="inputName"]').element.value).toBe(
      "carrot-test"
    );
    expect(wrapper.find('input[id="inputCostPerAmount"]').element.value).toBe(
      "1"
    );
    expect(wrapper.find('input[id="inputAmount"]').element.value).toBe("500");
    expect(wrapper.find('select[id="inputUnit"]').element.value).toBe("gram");

    expect(wrapper.vm.$data.ingredient).toEqual({
      article_number: "999",
      name: "carrot-test",
      cost_per_amount: "1",
      amount: "500",
      unit: "gram",
    });
  });

  it("clear button wipes loaded data out", async () => {
    await wrapper.setData({
      ingredient: {
        article_number: "999",
        name: "carrot-test",
        cost_per_amount: "1",
        amount: "500",
        unit: "gram",
      },
    });

    expect(wrapper.vm.$data.ingredient).toEqual({
      article_number: "999",
      name: "carrot-test",
      cost_per_amount: "1",
      amount: "500",
      unit: "gram",
    });

    let clearBtn = wrapper.find('button[id="clearBtn"]');
    expect(clearBtn.exists()).toBe(true);
    expect(clearBtn.is("button")).toBe(true);
    await clearBtn.trigger("click");

    expect(wrapper.vm.$data.ingredient).toEqual({
      article_number: null,
      name: null,
      cost_per_amount: null,
      amount: null,
      unit: "gram",
    });

    expect(wrapper.find('input[id="inputArticleNumber"]').element.value).toBe(
      ""
    );
    expect(wrapper.find('input[id="inputName"]').element.value).toBe("");
    expect(wrapper.find('input[id="inputCostPerAmount"]').element.value).toBe(
      ""
    );
    expect(wrapper.find('input[id="inputAmount"]').element.value).toBe("");
    expect(wrapper.find('select[id="inputUnit"]').element.value).toBe("gram");
  });

  it("opens confirmation modal when delete button is pressed", async () => {
    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(0);
    await wrapper.setData({
      ingredients: [
        {
          article_number: "998",
          name: "carrot-sample",
          cost_per_amount: "1.01",
          amount: "501",
          unit: "gram",
        },
      ],
    });

    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(1);
    let deleteBtn = wrapper.find('button[id="deleteBtn"]');
    expect(deleteBtn.exists()).toBe(true);
    expect(deleteBtn.is("button")).toBe(true);
    let modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(false);

    await deleteBtn.trigger("click");

    modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(true);
    expect(modal.is("div")).toBe(true);
    expect(modal.text()).toContain("Are you sure about that?");
    expect(modal.text()).toContain("Nevermind..");
    expect(modal.text()).toContain("Just do it!");
  });

  it("closes confirmation modal when cancel button is pressed", async () => {
    await wrapper.setData({
      ingredients: [
        {
          article_number: "998",
          name: "carrot-sample",
          cost_per_amount: "1.01",
          amount: "501",
          unit: "gram",
        },
      ],
    });
    let deleteBtn = wrapper.find('button[id="deleteBtn"]');
    expect(deleteBtn.exists()).toBe(true);
    expect(deleteBtn.is("button")).toBe(true);
    await deleteBtn.trigger("click");

    let cancelBtn = wrapper.find('button[id="cancelBtn"]');
    expect(cancelBtn.exists()).toBe(true);
    expect(cancelBtn.is("button")).toBe(true);
    await cancelBtn.trigger("click");

    let modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(false);
    cancelBtn = wrapper.find('button[id="cancelBtn"]');
    expect(cancelBtn.exists()).toBe(false);
  });

  it("clear search button wipes searched data out", async () => {
    let searchField = wrapper.find('input[id="searchField"]');
    expect(searchField.exists()).toBe(true);
    expect(searchField.is("input")).toBe(true);
    await searchField.setValue("sample ingredient");
    expect(searchField.element.value).toBe("sample ingredient");

    let clearSearchBtn = wrapper.find('button[id="clearSearchBtn"]');
    expect(clearSearchBtn.exists()).toBe(true);
    expect(clearSearchBtn.is("button")).toBe(true);
    await clearSearchBtn.trigger("click");
    expect(searchField.element.value).toBe("");
  });

  it("filters searched data", async () => {
    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(0);
    await wrapper.setData({
      ingredients: [
        {
          article_number: "998",
          name: "sample ingredient X",
          cost_per_amount: "1.01",
          amount: "501",
          unit: "kilogram",
        },
        {
          article_number: "997",
          name: "sample ingredient Y",
          cost_per_amount: "1.02",
          amount: "502",
          unit: "liter",
        },
      ],
    });
    let rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(2);

    // filters sample-X by name
    let searchField = wrapper.find('input[id="searchField"]');
    expect(searchField.exists()).toBe(true);
    expect(searchField.is("input")).toBe(true);
    await searchField.setValue("x");
    rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(1);
    expect(columnData(1, rows)[0]).toBe("sample ingredient X");

    // filters sample-Y by name
    await searchField.setValue("y");
    rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(1);
    expect(columnData(1, rows)[0]).toBe("sample ingredient Y");

    // filters sample-X by article_number
    await searchField.setValue("998");
    rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(1);
    expect(columnData(0, rows)[0]).toBe("998");
    expect(columnData(1, rows)[0]).toBe("sample ingredient X");

    // filters sample-Y by article_number
    await searchField.setValue("997");
    rows = wrapper.findAll("tbody > tr").wrappers;
    expect(rows).toHaveLength(1);
    expect(columnData(0, rows)[0]).toBe("997");
    expect(columnData(1, rows)[0]).toBe("sample ingredient Y");
  });
});
