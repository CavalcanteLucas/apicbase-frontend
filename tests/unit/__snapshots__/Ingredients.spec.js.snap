// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ingredients view renders 1`] = `
<div id="ingredients">
  <div class="container">
    <div class="col-12 mx-auto">
      <div class="row">
        <div class="col-md-4 mx-auto">
          <div class="input-group mb-3"><span id="search-ingredient" class="input-group-text"><img src="../assets/img/search.svg" alt="Search"></span> <input type="text" id="searchField" class="form-control"> <button type="button" id="clearSearchBtn" class="btn btn-outline-secondary"><img src="../assets/img/backspace.svg" alt="Clear Search"></button></div>
        </div>
      </div>
      <div class="row justify-content-center" style="padding-top: 1rem;">
        <div class="col-md-10 offset-md-1">
          <form id="ingredientForm">
            <div class="row g-1">
              <div class="col-md-1"><input type="number" id="inputArticleNumber" placeholder="#" min="1" required="required" class="form-control"></div>
              <div class="col-md-3"><input type="text" id="inputName" placeholder="Name" aria-label="Name" required="required" class="form-control"></div>
              <div class="col-md-2"><input type="number" id="inputAmount" placeholder="Amount" min=".01" step=".01" required="required" class="form-control"></div>
              <div class="col-md-1"><select aria-label="Select unit" id="inputUnit" required="required" class="form-select">
                  <option value="unit">Unit</option>
                  <option value="gram">gram</option>
                  <option value="kilogram">kilogram</option>
                  <option value="liter">liter</option>
                  <option value="centiliter">centiliter</option>
                </select></div>
              <div class="col-md-2">
                <div class="input-group mb-3"><span id="inputCostPerAmount" class="input-group-text">â‚¬</span> <input type="number" id="inputCostPerAmount" placeholder="Cost" min=".01" step=".01" required="required" class="form-control"></div>
              </div>
              <div class="col-md-3"><button type="button" id="clearBtn" class="btn btn-outline-secondary float-start"><img src="../assets/img/backspace.svg" alt="Clear"></button> <button type="submit" id="submitBtn" class="btn btn-primary float-start" style="margin-left: 0.25rem;">
                  Save
                </button></div>
            </div>
          </form>
        </div>
      </div>
      <!---->
      <div class="row" style="padding-top: 2rem;">
        <div class="col-12 mx-auto">
          <table aria-label="Ingredients table" class="table table-striped table-hovering">
            <thead>
              <th scope="col"><strong>#</strong></th>
              <th scope="col">Name</th>
              <th scope="col">Amount</th>
              <th scope="col">Unit</th>
              <th scope="col">Cost</th>
            </thead>
            <tbody></tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
